shell/platform/android/io/flutter/plugin/common/StringMessageCodec.java

```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugin.common;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * A {@link MessageCodec} using UTF-8 encoded String messages.
+ */
+public final class StringMessageCodec implements MessageCodec<String> {
+    // This codec must match the Dart codec of the same name in package flutter/services.
+    public static final StringMessageCodec INSTANCE = new StringMessageCodec();
+
+    private StringMessageCodec() {
+    }
+
+    @Override
+    public ByteBuffer encodeMessage(String message) {
+        if (message == null) {
+            return null;
+        }
+        // TODO(mravn): Avoid the extra copy below.
+        final byte[] bytes = message.getBytes(StandardCharsets.UTF_8);
```

Comment:
There's something goofy in the way Java encodes non-BMP code points in UTF-8.  Maybe I'm misremembering?  It might be worth testing to make sure something like U+1F602 (😂) gets encoded and decoded properly.